### ISSUE TITLE
`code_change` from Phoenix Channel is not called on hot upgrade

### ISSUE THREAD

**Issue by pdawczak** [bitwalker/distillery#301]
It is linked to the [other issue](https://github.com/bitwalker/distillery/issues/300) where I was demystifying how `code_change` works and I couldn't manage to execute [`code_change`](https://hexdocs.pm/phoenix/Phoenix.Channel.html#c:code_change/3) for Phoenix Channel.

Please, find the whole code in [this repository](https://github.com/pdawczak/deploy_fun)

### Steps to reproduce

Please note, in here, I've defined a function that should always be called accepting any arguments when performing hot code upgrade: [code](https://github.com/pdawczak/deploy_fun/blob/master/lib/deploy_web/channels/fun_channel.ex#L38-L43)

There are two very similar functions ([handling `:after_join`](https://github.com/pdawczak/deploy_fun/blob/master/lib/deploy_web/channels/fun_channel.ex#L14-L23) and [other one](https://github.com/pdawczak/deploy_fun/blob/master/lib/deploy_web/channels/fun_channel.ex#L25-L36) called in different times of process lifecycle - I've used them for more verbose logging trying understand which pars of the code are executed.

When running my application (`PORT=4044 ./bin/deploy start`) and tailing the logs (`▶ tail -f var/log/erlang.log.1`) I was able to read all the logs when performing upgrade in other screen (like, for _plain GenServer_). The logs from `code_change` have never appeared.

I was able to confirm that the `code_change` for my [`FunChannel`](https://github.com/pdawczak/deploy_fun/blob/master/lib/deploy_web/channels/fun_channel.ex) has properly appeared in my compiled release:

```
▶ cd release/<latest_release>
▶ ack code_change
```

with result:

```
relup
21:    {code_change,up,[{'Elixir.Deploy.SampleServer',[]}]},
27:    {code_change,up,[{'Elixir.DeployWeb.FunChannel',[]}]},
55:    {code_change,down,[{'Elixir.Deploy.SampleServer',[]}]},
61:    {code_change,down,[{'Elixir.DeployWeb.FunChannel',[]}]},
```

### Description of issue

- What are the expected results?
    **the `code_change` callback should be executed on hot code upgrades**
- What OS, Erlang/Elixir versions are you seeing this issue on?
    * macOS Sierra 10.12.6
    * Erlang 20
    * Elixir 1.5
    * Phoenix 1.3
    * Distillery 1.4:

```
* distillery 1.4.1 (Hex package) (mix)
  locked at 1.4.1 (distillery) 546d851b
```

- If possible, also provide your `rel/config.exs`, as it is often
  my first troubleshooting question, and you'll save us both time :)
  **Default one generated by `mix release.init` and example of it can be found [here](https://github.com/pdawczak/deploy_fun/blob/master/rel/config.exs)**
- Is there documentation that says one thing, but Distillery does
  another? If so, please link the doc here so it can be updated if
  it's a documentation issue, or so that the fix can be based around
  what's documented.
**There is no documentation explaining how it suppose to work in [original docs](https://hexdocs.pm/phoenix/Phoenix.Channel.html#c:code_change/3), but it looks very similar to the definition from [`GenServer.code_change/3`](https://hexdocs.pm/elixir/GenServer.html#c:code_change/3)**

Thanks for your help looking into this!

**Comment by bitwalker** [comment]
So the reason why this isn't working as you are expecting is because the Channel module is not actually run as a process - instead it is a callback module which is delegated to from `Phoenix.Channel.Server` and is stored in the state of the socket. What this means in practice is that your `code_change` handler will only get called if there is a change to the `Phoenix.Channel.Server` module in Phoenix which triggers an upgrade. What will happen during the upgrade is once the `FunChannel` code has been updated by the release handler, any future calls to it from `Phoenix.Channel.Server` will use the new version of the code, rather than the old version. This occurs because Phoenix is using fully-qualified functions calls, i.e. `module.function(...)`.

For named processes, you can add `{apply, {m, f, a}}` instructions to the generated appup file (it will be generated to `_build/<env>/rel/<app>/lib/<app>-<new_version>/ebin/<app>.appup`, to use your modifications, you need to save it to `_build/<env>/lib/<app>/ebin/<app>.appup`. The issue with how Phoenix channels are set up is that they do not have a registered name, so there is not an easy way to get the pids of all of the sockets, and if there were, you'd still have to define functions in your module which could be called via `apply` and handle looking up the pids and then calling `:sys.change_code/4`.

TL;DR - Your code is already being properly upgraded, just not the way you would expect due to how Phoenix channels are implemented.

**Comment by michalmuskala** [comment]
So the matter is a bit more complex. You can have a process that wraps a GenServer and it will be correctly handled during upgrades, for example with the `connection` library.

The problem with phoenix channels is a different one. The channel processes are not started under a supervisor. During an upgrade the supervision tree is inspected and only processes in the tree are upgraded, but channels are started directly by the transport process without a supervisor. This means that the upgrade mechanism has no way to reach the channel processes.

I consider this to be a small flaw in the phoenix design and maybe something to address in the future. Otherwise phoenix should clearly state that upgrades are not supported and you need custom `appups` as @bitwalker described to achieve that.

\cc @chrismccord

**Comment by bitwalker** [comment]
@michalmuskala I was looking through the code, but I didn't see anywhere that the module implementing `Phoenix.Channel` is actually backed by a process itself (granted I was in a hurry). I did see where the general `Phoenix.Channel.Server` process calls out to `socket.channel.code_change(...)` or something to that effect, but my takeaway was that these modules are callback modules, not actually individually backed by a process. I may have overlooked something though.

Good point about the fact that processes are only upgraded if they are in the supervision tree - I assumed they were in this case, but that's never a safe assumption to make, good catch!

**Comment by bitwalker** [comment]
On the more general topic of whether Phoenix should be amenable to upgrades - I'm of the opinion that while Phoenix doesn't need to go out of it's way to make the support flawless, it should probably support the fundamental capability to be hot upgraded. If Phoenix decides *not* to support hot upgrades explicitly, that can be handled in appups by just having the release handler restart Phoenix, but it will essentially defeat the purpose of using hot upgrades if the core of your application has to go down (even if for only a short time). 

I know there are quite a few teams using Phoenix and hot upgrades, and while there have been hiccups, the basic ability to use hot upgrades has thus far not been an issue - I think it's worth investing the effort to make it compatible, particularly since Phoenix is basically the most popular framework in the community - to be explicitly incompatible with hot upgrades would send a signal that hot upgrades are to be avoided, which I don't think is the right message. Caution is warranted yes, and documenting that Phoenix may not support hot upgrades flawlessly should warn people who aren't dead serious about using them to steer clear, unless they really do need the capability.

That's my two cents anyway, but I'd definitely like to know which way the team decides to go, because I'm sure I will get plenty of questions about it ;)

**Comment by michalmuskala** [comment]
Reported to phoenix https://github.com/phoenixframework/phoenix/issues/2432

**Comment by pdawczak** [comment]
Great! Thanks @michalmuskala. I was going to do this over the weekend when I had a moment to put the issue together

**Comment by pdawczak** [comment]
Hey @bitwalker, thanks for suggestion - I was trying to apply it, but I turned to be a bit more difficult (not exactly clear). Is there article/tutorial you could suggest for reading that would expand on this? That would be great help!

**Comment by bitwalker** [comment]
@pdawczak Which suggestion in particular, and I'll do my best to explain it better :)

**Comment by pdawczak** [comment]
Thanks @bitwalker, I really appreciate it!

I've been checking the locations you've mentioned here:

> For named processes, you can add `{apply, {m, f, a}}` instructions to the generated appup file (it will be generated to `_build/<env>/rel/<app>/lib/<app>-<new_version>/ebin/<app>.appup`, to use your modifications, you need to save it to `_build/<env>/lib/<app>/ebin/<app>.appup`. The issue with how Phoenix channels are set up is that they do not have a registered name, so there is not an easy way to get the pids of all of the sockets, and if there were, you'd still have to define functions in your module which could be called via `apply` and handle looking up the pids and then calling `:sys.change_code/4`.

Unfortunately I'm not too familiar with the process of _application_. I did manage to find the location you've mentioned and the content of file looks like:

```
# deploy/_build/prod/rel/deploy/lib/deploy-0.0.22/ebin/deploy.appup

{"0.0.22",
 [{"0.0.21",
   [{load_module,'Elixir.Deploy.Repo',[]},
    {update,'Elixir.Deploy.SampleServer',{advanced,[]},[]},
    {load_module,'Elixir.DeployWeb.ErrorHelpers',[]},
    {load_module,'Elixir.DeployWeb.ErrorView',[]},
    {update,'Elixir.DeployWeb.FunChannel',{advanced,[]},[]},
    {load_module,'Elixir.DeployWeb.Gettext',[]},
    {load_module,'Elixir.DeployWeb.LayoutView',[]},
    {load_module,'Elixir.DeployWeb.PageController',[]},
    {load_module,'Elixir.DeployWeb.PageView',[]},
    {load_module,'Elixir.DeployWeb.UserSocket',[]},
    {load_module,'Elixir.DeployWeb.Router',
                 ['Elixir.DeployWeb.PageController']},
    {load_module,'Elixir.DeployWeb.Endpoint',['Elixir.DeployWeb.Router']},
    {load_module,'Elixir.Deploy.Application',['Elixir.DeployWeb.Endpoint']}]}],
 [{"0.0.21",
   [{load_module,'Elixir.Deploy.Repo',[]},
    {update,'Elixir.Deploy.SampleServer',{advanced,[]},[]},
    {load_module,'Elixir.DeployWeb.ErrorHelpers',[]},
    {load_module,'Elixir.DeployWeb.ErrorView',[]},
    {update,'Elixir.DeployWeb.FunChannel',{advanced,[]},[]},
    {load_module,'Elixir.DeployWeb.Gettext',[]},
    {load_module,'Elixir.DeployWeb.LayoutView',[]},
    {load_module,'Elixir.DeployWeb.PageController',[]},
    {load_module,'Elixir.DeployWeb.PageView',[]},
    {load_module,'Elixir.DeployWeb.UserSocket',[]},
    {load_module,'Elixir.DeployWeb.Router',
                 ['Elixir.DeployWeb.PageController']},
    {load_module,'Elixir.DeployWeb.Endpoint',['Elixir.DeployWeb.Router']},
    {load_module,'Elixir.Deploy.Application',
                 ['Elixir.DeployWeb.Endpoint']}]}]}.
```

Where I can see:

```
{update,'Elixir.DeployWeb.FunChannel',{advanced,[]},[]},
```

So I suspect, what I need to do, is to add the line you've mentioned like:

```
{apply,{'Elixir.DeployWeb.FunChannel',change_code,['0.0.2']}},
```

as per `{apply, {m, f, a}}`, but here are some questions - it looks like there are two lists that start with:

```
[{"0.0.21",
```

* Do I need to duplicate my entry in both of them?
* I also suspect the order of elements in lists is important, so where should I place my definition?
* Once I have this file prepared properly - how do I include it to my release so I can apply those changes in hot-code-upgrade manner? (which normally is moving my `tar` file into location where I'll run the `./bin/deploy upgrade <version>`).

Sorry for lengthy post! I really appreciate your help!

If the answer is a bit more complicated, I'll happily accept any resource from which I could learn, if you could suggest one!

**Comment by bitwalker** [comment]
Your example of how you will use `apply` is correct. You need to put the instruction in both lists, the first list is upgrading from the older to the newer version, and the second list is downgrading from the newer version to the older. You need to order the instructions accordingly, i.e. you need to use `apply` *after* the code for the module has been upgraded, and *before* it gets downgraded.

You use the `.appup` you created by placing it in `_build/<env>/lib/<app>/ebin/<app>.appup`, and then run `mix release`, it will be picked up and used as the appup for the release - but keep it somewhere else so that you don't lose the file if you happen to run `mix clean` or something. I'm tracking an issue to make working with appup files easier, and that should be available in the next release.

Hope that helps :)
